### PR TITLE
Feature/gsye 434

### PR DESCRIPTION
**Pull request by hannesdiedrich**
## Reason for the proposed changes

The SCM strategies should be able to receive forecasts and measurements from the gsy-e-sdk

## Proposed changes

- Add ability to digest forecast and measurements sent by the gsy-e-sdk 
- fix pylint warning (you better review this commit separately) 

**INTEGRATION_TESTS_BRANCH**=feature/GSYE-434
**GSY_FRAMEWORK_BRANCH**=master

### CODE REVIEW

File: src/gsy_e/gsy_e_core/sim_results/file_export_endpoints.py [hunk 1]
```diff
@@ -335,22 +357,24 @@ def rows(self) -> List:
 
 
 class CoefficientLeafDataExporter(BaseDataExporter):
+    """LeafDataExporter for SCm simulations."""
```
Comment (spyrostz) [inline src/gsy_e/gsy_e_core/sim_results/file_export_endpoints.py:360]: Minor typo, SCM with uppercase

File: src/gsy_e/gsy_e_core/simulation/results_manager.py [hunk 2]
```diff
@@ -157,20 +157,20 @@ def update_scm_manager(self, scm_manager: SCMManager) -> None:
...
         """Update the csv results on market cycle."""
         if self.export_results_on_finish:
             self._export.data_to_csv(area, current_time_slot, slot_no == 0, scm_manager)
 
-    def save_csv_results(self, area: "Area") -> None:
+    def save_csv_results(self, area: "CoefficientArea") -> None:
         """Update the CSV results on finish, and write the CSV files."""
         if self.export_results_on_finish:
             log.info("Exporting simulation data.")
-            self._export.data_to_csv(area, False, None)
```
Comment (spyrostz) [inline src/gsy_e/gsy_e_core/simulation/results_manager.py:173]: How come? Was it due to an error on the generation of the CSV files ?

Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/simulation/results_manager.py:173]: yes, the last market slot was written twice to the CSV files.

File: src/gsy_e/gsy_e_core/simulation/simulation.py [hunk 3]
```diff
@@ -438,10 +448,24 @@ def _execute_simulation(
             self.progress_info.update(slot_no, slot_count, self._time, self.config)
 
             self.area.cycle_coefficients_trading(self.progress_info.current_slot_time)
+            log.error(self.area.current_market_time_slot)
```
Comment (spyrostz) [inline src/gsy_e/gsy_e_core/simulation/simulation.py:451]: I think that this can be removed, right ?

File: src/gsy_e/gsy_e_core/simulation/simulation.py [hunk 4]
```diff
@@ -438,10 +448,24 @@ def _execute_simulation(
...
 
+            if self.config.external_connection_enabled:
+
+                self.config.external_redis_communicator.approve_aggregator_commands()
+
+                self.area.market_cycle_external()
+
+                self.config.external_redis_communicator.\
+                    publish_aggregator_commands_responses_events()
+
```
Comment (spyrostz) [inline src/gsy_e/gsy_e_core/simulation/simulation.py:468]: Hmmm, I do not like it that much that external connection handling is in the main simulation method. Can these if blocks not reside in a separate class? For instance, I think that all these can be wrapped around a class that we will only have to call once and execute all these methods in the class body instead of the _execute_simulation.

File: src/gsy_e/gsy_e_core/simulation/simulation.py [hunk 5]
```diff
@@ -422,14 +425,21 @@ def _time_since_start(self) -> Duration:
...
     def _execute_simulation(
             self, slot_resume: int, _tick_resume: int, console: NonBlockingConsole = None) -> None:
         slot_count, slot_resume = (
             self._time.calc_resume_slot_and_count_realtime(
                 self.config, slot_resume))
 
+        if self.config.external_connection_enabled:
+            self.config.external_redis_communicator.sub_to_aggregator()
+            self.config.external_redis_communicator.start_communication()
+
```
Comment (spyrostz) [inline src/gsy_e/gsy_e_core/simulation/simulation.py:442]: I think that this is also not needed, if these were performed in the constructor of the external_redis_communicator.

File: src/gsy_e/gsy_e_core/simulation/simulation.py [hunk 6]
```diff
@@ -411,6 +411,9 @@ def _init(self) -> None:
 
         self.area.activate_energy_parameters(self.config.start_date)
 
+        if self.config.external_connection_enabled:
+            global_objects.scm_external_global_stats(self.area)
```
Comment (spyrostz) [inline src/gsy_e/gsy_e_core/simulation/simulation.py:415]: Same comment as below, I think that it is better to embed the if block inside the body of the `scm_external_global_stats.__call__` in order to not have the if block in the simulation class

File: src/gsy_e/gsy_e_core/global_objects_singleton.py [hunk 7]
```diff
@@ -26,6 +27,7 @@ class GlobalObjects:
     profiles_handler = ProfilesHandler()
     external_global_stats = ExternalConnectionGlobalStatistics()
     future_market_counter = FutureMarketCounter()
+    scm_external_global_stats = SCMExternalConnectionGlobalStatistics()
```
Comment (spyrostz) [inline src/gsy_e/gsy_e_core/global_objects_singleton.py:30]: I think that the singleton can be avoided by dependency injection on the simulation class. Essentially the simulation class could potentially keep a member with an instance of this class, and pass it to the classes that need it. The complex part is that external_redis_communicator (that requires this) is part of SimulationConfig, therefore it would also need to be passed in the SimulationConfig class. However I think that it would be much better WRT code grouping than the singleton, and it can be avoided

**Review by spyrostz — COMMENTED** [review]
Some comments, most about code structure. Functionality-wise it looks good!

File: src/gsy_e/gsy_e_core/global_objects_singleton.py (hunk 7 above)
Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/global_objects_singleton.py:30]: This comment caught me off guard. I remember discussion in length how to make the `grid_stats_tree` available in multiple places of the code and decided for a singleton. Now I spend some time to investigate and indeed it turns out, that we only need the grid_stats_tree only here, as you said: 
`Simulation->SimulationConfig->ExternalConnectionCommunicator->AggregatorHandler`
However, in order to create and update the `grid_stats_tree`, we need to pass the root area like done [here](https://github.com/gridsingularity/gsy-e/blob/36d8a6fe0fe93b69fd5b88ddb44281c1de3d9699/src/gsy_e/gsy_e_core/simulation/simulation.py#L112). 
The class will probably need an `update_area` method then. Let me try this (just fearing that I already tried this in the past which was not successful and lead to the singleton)

File: src/gsy_e/gsy_e_core/simulation/simulation.py (hunk 5 above)
Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/simulation/simulation.py:442]: Let me try this. I only copied this from the main simulation class. There must be a reason why we did not put these in the constructor of the `ExternalConnectionCommunicator` class.

File: src/gsy_e/gsy_e_core/simulation/simulation.py (hunk 4 above)
Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/simulation/simulation.py:468]: I played around a lot with the order of these (also mixed them up with the the other commands in this method). After finding the right order i wanted to refactor this into a separate method, but forgot. I agree, lets even add a class that deals with the external connections on the SCM simulations.

File: src/gsy_e/gsy_e_core/simulation/simulation.py (hunk 3 above)
Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/simulation/simulation.py:451]: yes, remnant of testing (normally I use print() which is why this got overlooked).

File: src/gsy_e/gsy_e_core/global_objects_singleton.py (hunk 7 above)
Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/global_objects_singleton.py:30]: @spyrostz I made good progress on this, but unfortunately, I found out, that we also need access to the `ExternalConnectionGlobalStatistics` in the external mixing [here](https://github.com/gridsingularity/gsy-e/blob/36d8a6fe0fe93b69fd5b88ddb44281c1de3d9699/src/gsy_e/models/strategy/external_strategies/__init__.py#L472) and [here](https://github.com/gridsingularity/gsy-e/blob/36d8a6fe0fe93b69fd5b88ddb44281c1de3d9699/src/gsy_e/models/strategy/external_strategies/__init__.py#L542) [whereas the second usage seems to be avoidable]
The only solution I see is to create an `EXTERNAL_TICK` event and to controll the dispatching from the simulation class, which I think is an overkill. What do you think.

File: src/gsy_e/gsy_e_core/simulation/simulation.py (hunk 4 above)
Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/simulation/simulation.py]: As the external communication is done a bit differently in the normal simulation I decided to only move these steps into a separate method in the SCMSimulation: [0d2cbfc](https://github.com/gridsingularity/gsy-e/pull/1559/commits/0d2cbfcb410817f89992821aed22687ae80b80af)
Please let me know what you think. Thanks

File: src/gsy_e/gsy_e_core/simulation/simulation.py (hunk 3 above)
Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/simulation/simulation.py]: I fixed this in here [0d2cbfc](https://github.com/gridsingularity/gsy-e/pull/1559/commits/0d2cbfcb410817f89992821aed22687ae80b80af)

File: src/gsy_e/gsy_e_core/simulation/simulation.py [hunk 8]
```diff
@@ -430,16 +427,26 @@ def _deactivate_areas(self, area: "AreaBase"):
         for child in area.children:
             self._deactivate_areas(child)
 
+    def _handle_external_communication(self):
```
Comment (spyrostz) [inline src/gsy_e/gsy_e_core/simulation/simulation.py:430]: I like a lot that the external connection stuff are at least extracted to their own method! However, is it possible to not even call these methods using an early-return? For instance:

```suggestion
    def _handle_external_communication(self):
        if not self.config.external_connection_enabled:
            return
```
Would that work out ?

**Review by spyrostz — DISMISSED** [review]
Looks much better now, and left the LGTM already even though I have one minor comment! 

As a general refactoring step for the future though (and frankly the root cause of my former comments), I have to suggest to take out the external_redis_communicator from SimulationConfig class. I know that this is a long-standing issue and we have discussed it again in the past, however I believe that this will become even less manageable in the future if we add more external connections (in the sense that we will be obliged to use more singletons).
To summarise my thoughts, if the external_redis_communicator was not hidden inside config, then we could pass the area object to it and encapsulate this whole logic inside the communicator, without having singletons and it being completely decoupled. This is outside the scope of this PR though, and I will probably open a ticket in order to explicitly refactor this. Thanks !

File: src/gsy_e/gsy_e_core/simulation/simulation.py (hunk 8 above)
Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/simulation/simulation.py:430]: That should work. All the methods inside check again if the aggregator is enabled, but why not returning early. Will do. Thanks!

**Review by spyrostz — DISMISSED** [review]

File: src/gsy_e/gsy_e_core/simulation/simulation.py (hunk 5 above)
Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/simulation/simulation.py]: @spyrostz we have a problem with this change. The connection to redis might not be initiated in this when calling these methods in the init of the external_redis_communicator [here](https://github.com/gridsingularity/gsy-e/pull/1559/commits/0d2cbfcb410817f89992821aed22687ae80b80af#diff-dacdd5512566d09ead79d7866dc73d81fffa3907abad2f0880ddb06c8af5d343R158) (have a look at the failing tests, you can reproduce these locally by turning off the redis server)
This is probably why it was pulled out of there into the simulation class. I propose to add an `activate` method that is called after the init (from the simulation class) which is the same implementation as before, only grouped into a method. What do you thin?

Comment (spyrostz) [reply src/gsy_e/gsy_e_core/simulation/simulation.py]: @hannesdiedrich too bad to hear! I agree with your suggestion, and I think that it is good because it will separate the parts of the code that need to be part of the activate process. Thanks again!

Comment (hannesdiedrich) [reply src/gsy_e/gsy_e_core/simulation/simulation.py]: I did that [cee9417](https://github.com/gridsingularity/gsy-e/pull/1559/commits/cee9417511d93ec337c3e32943889dc3aa19c5a4)
